src/jdk.incubator.vector/share/classes/jdk/incubator/vector/IntVector.java

```diff
@@ -591,16 +593,66 @@ IntVector lanewise(VectorOperators.Unary op,
      * @see #lanewise(VectorOperators.Binary,int,VectorMask)
      */
     @Override
-    public abstract
+    @ForceInline
+    public final
     IntVector lanewise(VectorOperators.Binary op,
-                                  Vector<Integer> v);
+                                  Vector<Integer> v) {
+        IntVector that = (IntVector) v;
+        that.check(this);
+        if (op == ROR || op == ROL) { // FIXME: JIT should do this
+            IntVector neg = that.lanewise(NEG);
+            IntVector hi = this.lanewise(LSHL, (op == ROR) ? neg : that);
+            IntVector lo = this.lanewise(LSHR, (op == ROR) ? that : neg);
+            return hi.lanewise(OR, lo);
+        }
+
+        if (op == DIV) {
+            VectorMask<Integer> eqz = that.eq((int)0);
+            if (eqz.anyTrue()) {
+                throw that.divZeroException();
+            }
+        }
+        return lanewise0(op, that, null);
+    }
+
+    /**
+     * {@inheritDoc} <!--workaround-->
+     * @see #lanewise(VectorOperators.Binary,int,VectorMask)
+     */
+    @Override
     @ForceInline
-    final
-    IntVector lanewiseTemplate(VectorOperators.Binary op,
-                                          Vector<Integer> v) {
+    public final
+    IntVector lanewise(VectorOperators.Binary op,
+                                  Vector<Integer> v,
+                                  VectorMask<Integer> m) {
+        if (op == ROR || op == ROL) {
+            return blend(lanewise(op, v), m);
+        }
+
         IntVector that = (IntVector) v;
         that.check(this);
-        if (opKind(op, VO_SPECIAL  | VO_SHIFT)) {
+        if (op == DIV) {
+            VectorMask<Integer> eqz = that.eq((int)0);
+            if (eqz.and(m).anyTrue()) {
+                throw that.divZeroException();
+            }
+            // suppress div/0 exceptions in unset lanes
+            that = that.lanewise(NOT, eqz);
```

Comment:
I guess this is required because we don't know how the intrinsic will support masking? How can a  div/0 exception can occur?

Reply:
Thanks for looking at this PR @PaulSandoz ! Yes, this is actually required for masked division operation. I think it has to make sure no div/0 exception happens during hotspot compiling the intrinsic? Am I missing something? Thanks!

Reply:
Did you observe any issues without this? I understand how such exceptions can occur when compiling the instrinsic, but could be missing something.

Reply:
I tried to remove `that = that.lanewise(NOT, eqz)`, but not issues are found. I think this is because currently the hotspot doesn't implement the intrinsic for integer division. And it seems this is only added for the division with integer types which might not be useful. And there are no need with any zero checks for `float/double` types in JAVA.